File: chapitre5/exercice5.py
import threading
import time
import random
import sys

compteur = 0
mon_mutex = threading.Lock()

def incrementation(limite):
    global compteur
    while True:
        with mon_mutex:
            if compteur >= limite:
                break
            compteur += 1
        time.sleep(random.randint(1, 5))

def affiche(limite):
    global compteur
    while True:
        with mon_mutex:
            if compteur >= limite:
                break
            print(f"Valeur actuelle du compteur : {compteur}")
        time.sleep(2)

def main():
    global compteur
    if len(sys.argv) < 2:
        return 1
    limite = int(sys.argv[1])

    t1 = threading.Thread(target=incrementation, args=(limite,))
    t2 = threading.Thread(target=affiche, args=(limite,))

    t1.start()
    t2.start()

    t1.join()
    t2.join()

    print(f"Fin du programme. Compteur final : {compteur}")
    return 0

if __name__ == "__main__":
    sys.exit(main())

File: chapitre5/exercice5.c
#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>

int compteur = 0; 
pthread_mutex_t mon_mutex = PTHREAD_MUTEX_INITIALIZER;

void* incrementation(void* arg) {
    int limite = (int)(long)arg;
    while (1) {
        pthread_mutex_lock(&mon_mutex);
        if (compteur >= limite) {
            pthread_mutex_unlock(&mon_mutex);
            break;
        }
        compteur++; 
        pthread_mutex_unlock(&mon_mutex);
        
        sleep(rand() % 5 + 1);
    }
    pthread_exit(NULL); 
}

void* affiche(void* arg) {
    int limite = (int)(long)arg;
    while (1) {
        pthread_mutex_lock(&mon_mutex);
        if (compteur >= limite) {
            pthread_mutex_unlock(&mon_mutex);
            break;
        }
        printf("Valeur actuelle du compteur : %d\n", compteur);
        pthread_mutex_unlock(&mon_mutex);
        
        sleep(2); // Affichage toutes les 2 secondes [1]
    }
    pthread_exit(NULL);
}

int main(int argc, char *argv[]) {
    if (argc < 2) return 1;
    int limite = atoi(argv[1]);
    pthread_t t1, t2;
    srand(time(NULL));

    pthread_create(&t1, NULL, incrementation, (void*)(long)limite);
    pthread_create(&t2, NULL, affiche, (void*)(long)limite);

    pthread_join(t1, NULL);
    pthread_join(t2, NULL);

    printf("Fin du programme. Compteur final : %d\n", compteur);
    return 0;
}
